consistently failing tests on some architectures, flaky on others

On i386 the following fail: test_archive_mount, test_archive_extract, test_archive_diff, test_borg_prune, test_repo_unlink, test_repo_add_success, test_create, test_scheduler_create_backup

On armhf (32bit, used on the Raspberry pi 3, I think): same.

These tests also occasionally fail on Intel/AMD 64bit CI, and I've received release critical bugs about this issue (Debian bugtracker) for arm64 (used on the Raspberri pi ≥ 4, I think).  Flakiness on arm64 seems like it will affect Borg on laptops using Apple M1 chip, along with future ARM laptops, and I also wonder if it's useful fuzzing for amd64 (aka: x86-64).

Vorta for Debian is currently based on 0.7.1, plus the following patches (mostly cherry picked from master):
https://salsa.debian.org/python-team/packages/vorta/-/tree/debian/master/debian/patches

I've also increased all timeouts to 10 (where they were previously << 10) or 20  (where they were previously ≥10) seconds.  For the moment, this patch appears to have mitigated the rare nondeterministic test failures on amd64, and it remains to be seen whether it mitigates future failures on arm64.

The most interesting error output from `tests.reproducible-builds.org` is
```
ERROR    vorta.scheduler:scheduler.py:118 Conditions for backup not met. Aborting.
ERROR    vorta.scheduler:scheduler.py:119 Backup is already in progress.
```
and my primary concern is this might indicate the existence of some corner-case failures in Vorta's backup scheduling, possibly due to race conditions or missing locking conditions.

Logs include a list of the full stack of all installed dependencies, and the actual commands used to initiate tests.

https://tests.reproducible-builds.org/debian/rb-pkg/unstable/i386/vorta.html
https://tests.reproducible-builds.org/debian/rb-pkg/unstable/armhf/vorta.html

and here are some logs from another one of our three CI networks:
https://ci.debian.net/data/autopkgtest/testing/arm64/v/vorta/8520699/log.gz
https://ci.debian.net/data/autopkgtest/testing/i386/v/vorta/8521042/log.gz

These logs were cited in Debian [Bug #976134](https://bugs.debian.org/976134)

It's of course also possible that this is a bug in another Python package, or in Qt, or in Borg.  If I've made a mistake anywhere, please let me know, and I'll humbly accept responsibility and fix it :-)

Merry Christmas and Happy New Year!
Nicholas

@m3nu, please reopen this issue; it's unfortunately not fixed.  Both 0.7.2 and 52233a9 contain the following regression compared to 0.7.1:

```
=================================== FAILURES ===================================
_____________________________ test_archive_extract _____________________________

qapp = <vorta.application.VortaApp object at 0x7fa772488e50>
qtbot = <pytestqt.qtbot.QtBot object at 0x7fa77038c490>
mocker = <pytest_mock.MockFixture object at 0x7fa770392ee0>
borg_json_output = <function borg_json_output.<locals>._read_json at 0x7fa7703bd160>

    def test_archive_extract(qapp, qtbot, mocker, borg_json_output):
        main = qapp.main_window
        tab = main.archiveTab
        main.tabWidget.setCurrentIndex(3)
    
        tab.populate_from_profile()
        qtbot.waitUntil(lambda: tab.archiveTable.rowCount() == 2)
    
        tab.archiveTable.selectRow(0)
        stdout, stderr = borg_json_output('list_archive')
        popen_result = mocker.MagicMock(stdout=stdout, stderr=stderr, returncode=0)
        mocker.patch.object(vorta.borg.borg_thread, 'Popen', return_value=popen_result)
        qtbot.mouseClick(tab.extractButton, QtCore.Qt.LeftButton)
    
        qtbot.waitUntil(lambda: hasattr(tab, '_window'), **pytest._wait_defaults)
>       qtbot.waitUntil(lambda: tab._window == qapp.activeWindow(), **pytest._wait_defaults)
E       AssertionError: waitUntil timed out in 20000 miliseconds

tests/test_archives.py:128: AssertionError
----------------------------- Captured Qt messages -----------------------------
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
______________________________ test_archive_diff_______________________________

qapp = <vorta.application.VortaApp object at 0x7fa772488e50>
qtbot = <pytestqt.qtbot.QtBot object at 0x7fa7703a9b50>
mocker = <pytest_mock.MockFixture object at 0x7fa770366640>
borg_json_output = <function borg_json_output.<locals>._read_json at 0x7fa7683eb670>

    def test_archive_diff(qapp, qtbot, mocker, borg_json_output):
        main = qapp.main_window
        tab = main.archiveTab
        main.tabWidget.setCurrentIndex(3)
    
        tab.populate_from_profile()
        qtbot.waitUntil(lambda: tab.archiveTable.rowCount() == 2)
    
        stdout, stderr = borg_json_output('diff_archives')
        popen_result = mocker.MagicMock(stdout=stdout, stderr=stderr, returncode=0)
        mocker.patch.object(vorta.borg.borg_thread, 'Popen', return_value=popen_result)
    
        qtbot.mouseClick(tab.diffButton, QtCore.Qt.LeftButton)
        qtbot.waitUntil(lambda: hasattr(tab, '_window'), **pytest._wait_defaults)
>       qtbot.waitUntil(lambda: tab._window == qapp.activeWindow(), **pytest._wait_defaults)
E       AssertionError: waitUntil timed out in 20000 miliseconds

tests/test_archives.py:169: AssertionError
----------------------------- Captured Qt messages -----------------------------
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
    QtWarningMsg: QImage::scaleHeight: Image is a null image
=========================== short test summary info ============================
FAILED tests/test_archives.py::test_archive_extract - AssertionError: waitUnt...
FAILED tests/test_archives.py::test_archive_diff - AssertionError: waitUntil ...
=================== 2 failed, 49 passed, 2 skipped in 52.15s ===================
```
Full build log, including list of all installed dependencies available here:
[vorta_0.7.2+15.g52233a9-1_amd64-2021-02-17T04:40:47Z.build.txt](https://github.com/borgbase/vorta/files/5993386/vorta_0.7.2%2B15.g52233a9-1_amd64-2021-02-17T04.40.47Z.build.txt)

Debian 11 (bullseye) was frozen 12 Feb 2021, and one day was not enough time react to #788.  Debian 11 with currently ship with Vorta 0.7.1 due to this regression, and that release will be supported until somewhere around mid to late 2024, although thankfully most users will upgrade in 2023.  The window for applying for an exception to the freeze for one Vorta version upgrade will definitely close a week from now (24 Feb), so if that's something you're interested in, please fix whatever is causing test_archive_extract and test_archive_diff to fail.  I can help with bisecting over the weekend, but waiting for me will run the risk of cutting things too close.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -19,6 +19,19 @@ jobs:
       uses: actions/setup-python@v1
       with:
         python-version: ${{ matrix.python-version }}
+
+    - name: Get pip cache dir
+      id: pip-cache
+      run: |
+        echo "::set-output name=dir::$(pip cache dir)"
+    - name: pip cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.pip-cache.outputs.dir }}
+        key: ${{ runner.os }}-pip-${{ hashFiles('setup.cfg', 'requirements.d/**') }}
+        restore-keys: |
+          ${{ runner.os }}-pip-
+
     - name: Install system dependencies (Linux)
       if: runner.os == 'Linux'
       run: |
@@ -34,8 +47,10 @@ jobs:
       run: |
         pip install -e .
         pip install -r requirements.d/dev.txt
+
     # - name: Setup tmate session
     #   uses: mxschmitt/action-tmate@v1
+
     - name: Test with pytest (Linux)
       if: runner.os == 'Linux'
       run: |
@@ -45,6 +60,7 @@ jobs:
       if: runner.os == 'macOS'
       run: |
         pytest
+
     - name: Upload coverage to Codecov
       uses: codecov/codecov-action@v1
       env:
@@ -53,6 +69,7 @@ jobs:
       with:
         token: ${{ secrets.CODECOV_TOKEN }}
         env_vars: OS, python
+
   lint:
     runs-on: ubuntu-latest
     steps:

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -44,7 +44,7 @@ jobs:
     - name: Test with pytest (macOS)
       if: runner.os == 'macOS'
       run: |
-        pytest --cov=vorta
+        pytest
     - name: Upload coverage to Codecov
       uses: codecov/codecov-action@v1
       env:

--- a/src/vorta/application.py
+++ b/src/vorta/application.py
@@ -75,7 +75,7 @@ class VortaApp(QtSingleApplication):
         self.backup_cancelled_event.connect(self.backup_cancelled_event_response)
         self.message_received_event.connect(self.message_received_event_response)
         self.backup_log_event.connect(self.react_to_log)
-        self.aboutToQuit.connect(cleanup_db)
+        self.aboutToQuit.connect(self.quit_app_action)
         self.set_borg_details_action()
         self.installEventFilter(self)
 
@@ -100,6 +100,12 @@ class VortaApp(QtSingleApplication):
         if event.type() == QtCore.QEvent.ApplicationPaletteChange and source == self.tray.contextMenu():
             self.tray.set_tray_icon()
         return False
+
+    def quit_app_action(self):
+        del self.main_window
+        self.scheduler.shutdown()
+        self.backup_cancelled_event.emit()
+        cleanup_db()
 
     def create_backup_action(self, profile_id=None):
         if not profile_id:

--- a/src/vorta/utils.py
+++ b/src/vorta/utils.py
@@ -62,12 +62,13 @@ def get_directory_size(dir_path):
             if os.path.islink(file_path):
                 continue
 
-            stat = os.stat(file_path)
-
-            # Visit each file once
-            if stat.st_ino not in seen:
-                seen.add(stat.st_ino)
-                data_size += stat.st_size
+            try:
+                stat = os.stat(file_path)
+                if stat.st_ino not in seen:  # Visit each file only once
+                    seen.add(stat.st_ino)
+                    data_size += stat.st_size
+            except FileNotFoundError:
+                continue
 
     files_count = len(seen)
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -34,6 +34,7 @@ def qapp(tmpdir_factory):
     qapp = VortaApp([])  # Only init QApplication once to avoid segfaults while testing.
 
     yield qapp
+    qapp.quit()
 
 
 @pytest.fixture(scope='function', autouse=True)

--- a/tests/test_source.py
+++ b/tests/test_source.py
@@ -1,12 +1,8 @@
-import os
 import pytest
-import vorta.models
 import vorta.views
-from PyQt5.QtWidgets import QApplication
-from PyQt5 import QtCore
 
 
-def test_add_folder(qapp, qtbot, tmpdir, monkeypatch, choose_file_dialog):
+def test_add_folder(qapp, qtbot, mocker, monkeypatch, choose_file_dialog):
     monkeypatch.setattr(
         vorta.views.source_tab, "choose_file_dialog", choose_file_dialog
     )
@@ -17,7 +13,12 @@ def test_add_folder(qapp, qtbot, tmpdir, monkeypatch, choose_file_dialog):
     tab.sourceAddFolder.click()
     qtbot.waitUntil(lambda: tab.sourceFilesWidget.rowCount() == 2, **pytest._wait_defaults)
 
-    # Test paste button
-    QApplication.clipboard().setText(os.path.expanduser('~'))  # Load clipboard
-    qtbot.mouseClick(tab.paste, QtCore.Qt.LeftButton)
-    assert tab.sourceFilesWidget.rowCount() == 3
+    # Test paste button with mocked clipboard
+    mock_clipboard = mocker.Mock()
+    mock_clipboard.text.return_value = __file__
+    mocker.patch.object(vorta.views.source_tab.QApplication, 'clipboard', return_value=mock_clipboard)
+    tab.paste_text()
+    qtbot.waitUntil(lambda: tab.sourceFilesWidget.rowCount() == 3, **pytest._wait_defaults)
+
+    # Wait for directory sizing to finish
+    qtbot.waitUntil(lambda: len(qapp.main_window.sourceTab.updateThreads) == 0, **pytest._wait_defaults)

--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -2,7 +2,7 @@ import uuid
 from vorta.keyring.abc import VortaKeyring
 
 
-def test_keyring(qapp):
+def test_keyring():
     UNICODE_PW = 'kjalsdfüadsfäadsfß'
     REPO = f'vorta-test-repo.{uuid.uuid4()}.com:repo'  # Random repo URL
 

